crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,349 @@
+            NonZeroUsize::new(max_items).unwrap(),
+        )));
+        Self { inner, cache, _pd: PhantomData }
+    }
+
+    /// Puts a value into the cache, and returns the old value if it existed.
+    pub async fn put(&self, key: B256, value: String) -> TransportResult<Option<String>> {
+        let mut cache = self.cache.write().await;
+        Ok(cache.put(key, value))
+    }
+
+    /// Gets a value from the cache, if it exists.
+    pub async fn get(&self, key: &B256) -> TransportResult<Option<String>> {
+        // Need to acquire a write guard to change the order of keys in LRU cache.
+        let mut cache = self.cache.write().await;
+        let val = cache.get(key).cloned();
+        Ok(val)
+    }
+
+    /// Saves the cache to a file specified by the path.
+    /// If the files does not exist, it creates one.
+    /// If the file exists, it overwrites it.
+    pub async fn save_cache(&self, path: PathBuf) -> TransportResult<()> {
+        let cache = self.cache.read().await;
+        let file = std::fs::File::create(path).map_err(TransportErrorKind::custom)?;
+
+        // Iterate over the cache and dump to the file.
+        let entries = cache
+            .iter()
+            .map(|(key, value)| FsCacheEntry { key: *key, value: value.clone() })
+            .collect::<Vec<_>>();
+        serde_json::to_writer(file, &entries).map_err(TransportErrorKind::custom)?;
+        Ok(())
+    }
+
+    /// Loads the cache from a file specified by the path.
+    /// If the file does not exist, it returns without error.
+    pub async fn load_cache(&self, path: PathBuf) -> TransportResult<()> {
+        if !path.exists() {
+            return Ok(());
+        };
+        let file = std::fs::File::open(path).map_err(TransportErrorKind::custom)?;
+        let file = BufReader::new(file);
+        let entries: Vec<FsCacheEntry> =
+            serde_json::from_reader(file).map_err(TransportErrorKind::custom)?;
+        let mut cache = self.cache.write().await;
+        for entry in entries {
+            cache.put(entry.key, entry.value);
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl<P, T> Provider<T> for CacheProvider<P, T>
+where
+    P: Provider<T>,
+    T: Transport + Clone,
+{
+    #[inline(always)]
+    fn root(&self) -> &RootProvider<T> {
+        self.inner.root()
+    }
+
+    async fn get_block_by_number(
+        &self,
+        number: BlockNumberOrTag,
+        hydrate: bool,
+    ) -> TransportResult<Option<Block>> {
+        let hash =
+            RequestType::GetBlockByNumber((number, hydrate)).params_hash(self.inner.client())?;
+
+        if let Some(block) = self.get(&hash).await? {
+            let block = serde_json::from_str(&block).map_err(TransportErrorKind::custom)?;
+            println!("Cache hit");
+            return Ok(Some(block));
+        }
```

Comment:
this has to be repeated for every endpoint that we cache.

this should be simpler, either with a helper function or a macro